sql/planner/projection.go

```diff
@@ -0,0 +1,209 @@
+package planner
+
+import (
+	"errors"
+
+	"github.com/genjidb/genji/database"
+	"github.com/genjidb/genji/document"
+	"github.com/genjidb/genji/sql/query"
+	"github.com/genjidb/genji/sql/query/expr"
+)
+
+// A ProjectionNode is a node that uses the given expressions to create a new document
+// for each document of the stream. Each expression can extract fields from the incoming
+// document, call functions, execute arithmetic operations. etc.
+type ProjectionNode struct {
+	node
+
+	Expressions []ResultField
+	tableName   string
+
+	cfg *database.TableConfig
+	tx  *database.Transaction
+}
+
+var _ outputNode = (*ProjectionNode)(nil)
+var _ operationNode = (*ProjectionNode)(nil)
+
+// NewProjectionNode creates a node that uses the given expressions to create a new document
+// for each document of the stream. Each expression can extract fields from the incoming
+// document, call functions, execute arithmetic operations. etc.
+func NewProjectionNode(n Node, expressions []ResultField, tableName string) Node {
+	return &ProjectionNode{
+		node: node{
+			op:   Projection,
+			left: n,
+		},
+		Expressions: expressions,
+		tableName:   tableName,
+	}
+}
+
+// IsEqual returns true if other is a *ProjectionNode and contains the
+// same information.
+func (n *ProjectionNode) IsEqual(other Node) bool {
+	if !n.node.IsEqual(other) {
+		return false
+	}
+
+	on := other.(*ProjectionNode)
+	if n.tableName != on.tableName {
+		return false
+	}
+
+	if len(n.Expressions) != len(on.Expressions) {
+		return false
+	}
+
+	for i := range n.Expressions {
+		switch t := n.Expressions[i].(type) {
+		case Wildcard:
+			if _, ok := on.Expressions[i].(Wildcard); !ok {
+				return false
+			}
+		case ResultFieldExpr:
+			rf, ok := on.Expressions[i].(ResultFieldExpr)
+			if !ok {
+				return false
+			}
+
+			if t.ExprName != rf.ExprName {
+				return false
+			}
+
+			if !expr.Equal(t.Expr, rf.Expr) {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// Bind database resources to this node.
+func (n *ProjectionNode) Bind(tx *database.Transaction, params []expr.Param) (err error) {
+	n.tx = tx
+	if n.tableName == "" {
+		return
```

Comment:
Could we project without a table? Maybe we need to return an error here?

Reply:
Yes we can, `SELECT 1 + 1` gives this `∏(1 + 1)`